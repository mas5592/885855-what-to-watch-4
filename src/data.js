export const promo = {
  title: `The Grand Budapest Hotel`,
  genre: `Drama`,
  date: `2014`
};
