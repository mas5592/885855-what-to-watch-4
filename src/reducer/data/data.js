import {extend} from '../../utils/utils.js';
import {normalizeFilmData, normalizeFilmsData} from '../../utils/utils.js';

const initialState = {
  promo: {},
  films: [],
  currentFilmComments: []
};

const ActionType = {
  LOAD_FILMS: `LOAD_FILMS`,
  LOAD_PROMO_FILM: `LOAD_PROMO_FILM`,
  LOAD_COMMENTS: `LOAD_COMMENTS`
};

const Operation = {
  loadFilms: () => (dispatch, getState, api) => {
    return api.get(`/films`).then((response) => {
      dispatch(ActionCreator.loadFilms(normalizeFilmsData(response.data)));
    });
  },
  loadPromo: () => (dispatch, getState, api) => {
    return api.get(`/films/promo`).then((response) => {
      dispatch(ActionCreator.loadPromo(normalizeFilmData(response.data)));
    });
  },
  loadComments: (filmId) => (dispatch, getState, api) => {
    return api.get(`/comments/${filmId}`).then((response) => {
      dispatch(ActionCreator.loadComments(response.data));
    });
  }
};

const ActionCreator = {
  loadFilms: (films) => ({
    type: ActionType.LOAD_FILMS,
    payload: films
  }),
  loadPromo: (film) => ({
    type: ActionType.LOAD_PROMO_FILM,
    payload: film
  }),
  loadComments: (comments) => ({
    type: ActionType.LOAD_COMMENTS,
    payload: comments
  })
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_FILMS:
      return extend(state, {
        films: action.payload
      });
    case ActionType.LOAD_PROMO_FILM:
      return extend(state, {
        promo: action.payload
      });
    case ActionType.LOAD_COMMENTS:
      return extend(state, {
        currentFilmComments: action.payload
      });
  }

  return state;
};

export {reducer, ActionType, ActionCreator, Operation};
