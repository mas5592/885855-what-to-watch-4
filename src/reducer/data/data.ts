import {extend} from '../../utils';
import {adaptFilm} from '../../adapters/film';
import {ActionCreator as AppStateActionCreator} from '../state/state';
import {FilmType, ReviewType} from '../../types';
import history from '../../history';

interface DataActionInterface {
  type?: string;
  payload?: FilmType | Array<FilmType> | Array<ReviewType> | boolean;
}

interface InitialStateInterface {
  card?: FilmType | {};
 films?: Array<FilmType> | [];
  filmReviews?: Array<ReviewType> | [];
  favoriteFilms?: Array<FilmType> | [];
  isLoading?: boolean;
  isLoadError?: boolean;
  isReviewSending?: boolean;
  isDispatchSuccessful?: boolean;
  isDispatchError?: boolean;
}

const initialState: InitialStateInterface = {
  card: {},
 films: [],
  filmReviews: [],
  favoriteFilms: [],
  isLoading: true,
  isLoadError: false,
  isReviewSending: false,
  isDispatchSuccessful: false,
  isDispatchError: false,
};

const ActionType = {
  CATCH_LOAD_ERROR: `CATCH_LOAD_ERROR`,
  CHECK_IS_DATA_SENDING: `CHECK_IS_DATA_SENDING`,
  CHECK_IS_DISPATCH_ERROR: `CHECK_IS_DISPATCH_ERROR`,
  CHECK_IS_DISPATCH_SUCCESSFUL: `CHECK_IS_DISPATCH_SUCCESSFUL`,
  CLEAR_SENDING_ERROR: `CLEAR_SENDING_ERROR`,
  FINISH_LOADING: `FINISH_LOADING`,
  LOAD_FAVORITE_FILMS: `LOAD_FAVORITE_FILMS`,
  LOAD_FILM_CARD: `LOAD_FILM_CARD`,
  LOAD_FILM_REVIEWS: `LOAD_FILM_REVIEWS`,
  LOAD_FILS: `LOAD_FILS`,
};

const ActionCreator = {
  loadFilmCard: (card) => {
    return {
      type: ActionType.LOAD_FILM_CARD,
      payload: card
    };
  },

  loadPromo: (films) => {
    return {
      type: ActionType.LOAD_FILS,
      payload:films,
    };
  },

  loadFilmReviews: (filmReviews) => {
    return {
      type: ActionType.LOAD_FILM_REVIEWS,
      payload: filmReviews,
    };
  },

  loadFavoriteFilms: (films) => ({
    type: ActionType.LOAD_FAVORITE_FILMS,
    payload: films,
  }),

  finishLoading: () => ({
    type: ActionType.FINISH_LOADING,
    payload: false,
  }),

  catchLoadError: () => ({
    type: ActionType.CATCH_LOAD_ERROR,
    payload: true,
  }),

  checkIsReviewSending: (isReviewSending) => ({
    type: ActionType.CHECK_IS_DATA_SENDING,
    payload: isReviewSending,
  }),

  checkIsDispatchSuccessful: (isDispatchSuccessful) => ({
    type: ActionType.CHECK_IS_DISPATCH_SUCCESSFUL,
    payload: isDispatchSuccessful,
  }),

  checkIsDispatchError: (isDispatchError) => ({
    type: ActionType.CHECK_IS_DISPATCH_ERROR,
    payload: isDispatchError,
  }),

  clearSendingError: () => ({
    type: ActionType.CLEAR_SENDING_ERROR,
    payload: false,
  }),
};

const Operations = {
  loadFilmCard: () => (dispatch, getState, api) => {
    return api.get(`/films/promo`)
      .then((response) => {
        dispatch(ActionCreator.loadFilmCard(adaptFilm(response.data)));
        dispatch(AppStateActionCreator.setActiveFilm(adaptFilm(response.data)));

      })
      .catch(() => {
        dispatch(ActionCreator.catchLoadError());
      });
  },

  loadPromo: () => (dispatch, getState, api) => {
    return api.get(`/films`)
      .then((response) => {
        const films = response.data.map((film) => adaptFilm(film));
        dispatch(ActionCreator.loadPromo(films));
        dispatch(ActionCreator.finishLoading());
      })
      .catch(() => {
        dispatch(ActionCreator.catchLoadError());
      });
  },

  loadFilmReviews: (filmId) => (dispatch, getState, api) => {
    return api.get(`/comments/${filmId}`)
      .then((response) => {
        dispatch(ActionCreator.loadFilmReviews(response.data));
      })
      .catch(() => {
        dispatch(ActionCreator.catchLoadError());
      });
  },

  pushReview: (filmId, review) => (dispatch, getState, api) => {
    dispatch(ActionCreator.checkIsReviewSending(true));
    return api.post(`/comments/${filmId}`, {
      rating: review.rating,
      comment: review.comment,
    })
    .then(() => {
      dispatch(ActionCreator.checkIsReviewSending(false));
      dispatch(ActionCreator.checkIsDispatchSuccessful(true));
      dispatch(ActionCreator.checkIsDispatchError(false));

      dispatch(Operations.loadFilmReviews(filmId));
      history.goBack();
    })
    .catch(() => {
      dispatch(ActionCreator.checkIsReviewSending(false));
      dispatch(ActionCreator.checkIsDispatchSuccessful(false));
      dispatch(ActionCreator.checkIsDispatchError(true));
    });
  },

  loadFavoriteFilms: () => (dispatch, getState, api) => {
    return api.get(`/favorite`)
    .then((response) => {
      if (response.data) {
        const favoriteFilms = response.data.map((favoriteFilm) => adaptFilm(favoriteFilm));
        dispatch(ActionCreator.loadFavoriteFilms(favoriteFilms));
      }
    })
    .catch(() => {
      dispatch(ActionCreator.catchLoadError());
    });
  },

  changeIsFilmFavorite: (filmId, isFavorite) => (dispatch, getState, api) => {
    dispatch(ActionCreator.checkIsReviewSending(true));
    return api.post(`/favorite/${filmId}/${isFavorite ? 1 : 0}`)
    .then(() => {
      dispatch(ActionCreator.checkIsReviewSending(false));
      dispatch(ActionCreator.checkIsDispatchSuccessful(true));
      dispatch(ActionCreator.checkIsDispatchError(false));
      dispatch(Operations.loadPromo());
      dispatch(Operations.loadFilmCard());
    })
    .catch(() => {
      dispatch(ActionCreator.checkIsReviewSending(false));
      dispatch(ActionCreator.checkIsDispatchSuccessful(false));
      dispatch(ActionCreator.checkIsDispatchError(true));
    });
  },
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.LOAD_FILM_CARD:
      return extend(state, {
        card: action.payload,
      });
    case ActionType.LOAD_FILS:
      return extend(state, {
       films: action.payload,
      });
    case ActionType.LOAD_FILM_REVIEWS:
      return extend(state, {
        filmReviews: action.payload,
      });
    case ActionType.LOAD_FAVORITE_FILMS:
      return extend(state, {
        favoriteFilms: action.payload,
      });
    case ActionType.FINISH_LOADING:
      return extend(state, {
        isLoading: action.payload,
      });
    case ActionType.CATCH_LOAD_ERROR:
      return extend(state, {
        isLoadError: action.payload,
      });
    case ActionType.CHECK_IS_DATA_SENDING:
      return extend(state, {
        isReviewSending: action.payload,
      });
    case ActionType.CHECK_IS_DISPATCH_SUCCESSFUL:
      return extend(state, {
        isDispatchSuccessful: action.payload,
      });
    case ActionType.CHECK_IS_DISPATCH_ERROR:
      return extend(state, {
        isDispatchError: action.payload,
      });
    case ActionType.CLEAR_SENDING_ERROR:
      return extend(state, {
        isDispatchError: action.payload,
      });

  }

  return state;
};

export {ActionType, ActionCreator, Operations, reducer, initialState};
