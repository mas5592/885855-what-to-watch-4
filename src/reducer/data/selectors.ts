import {createSelector} from 'reselect';
import NameSpace from '../name-space';
import {FILTER_ALL_GENRES} from '../../const';

export const getFilms = (state) => state[NameSpace.DATA].films;
export const getFilm = (state) => state[NameSpace.DATA].card;
export const getFilmReviews = (state) => state[NameSpace.DATA].filmReviews;

export const getFavoriteFilms = (state) => state[NameSpace.DATA].favoriteFilms;
export const getIsLoadError = (state) => state[NameSpace.DATA].isLoadError;
export const getIsReviewSending = (state) => state[NameSpace.DATA].isReviewSending;
export const getIsDispatchSuccessful = (state) => state[NameSpace.DATA].isDispatchSuccessful;
export const getIsDispatchError = (state) => state[NameSpace.DATA].isDispatchError;
export const getIsLoading = (state) => state[NameSpace.DATA].isLoading;

export const getFilmsGenres = createSelector(
    getFilms,
    (films) => {
      const genres = new Set(films.map((film) => film.genre));
      return [FILTER_ALL_GENRES, ...genres];
    }
);

