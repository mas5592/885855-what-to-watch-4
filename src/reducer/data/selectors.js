import {createSelector} from 'reselect';
import Namespace from '../namespace.js';
import {FILTER_ALL_GENRES} from '../../data.js';

export const loadFilms = (state) => {
  return state[Namespace.DATA].films;
};

export const loadPromo = (state) => {
  return state[Namespace.DATA].promo;
};

export const loadComments = (state) => {
  return state[Namespace.DATA].currentFilmComments;
};

const filterFilmsByGenre = (state) => {
  const films = state[Namespace.DATA].films;
  const showedFilms = state[Namespace.STATE].showedFilms;
  const genre = state[Namespace.STATE].genre;

  return genre === FILTER_ALL_GENRES
    ? films.slice(0, showedFilms)
    : films.filter((film) => film.genre === genre).slice(0, showedFilms);
};

export const loadFilmsByGenre = createSelector(
    (state) => state,
    filterFilmsByGenre
);
