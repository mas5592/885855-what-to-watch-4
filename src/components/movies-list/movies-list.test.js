import React from 'react';
import renderer from 'react-test-renderer';
import MoviesList from './movies-list.jsx';
import {films} from '../../data.js';

it(`Should render MoviesList component`, () => {
  const tree = renderer
    .create(
        <MoviesList
          films={films}
          onFilmCardClick={() => {}}
          onFilmCardMouseOver={() => {}}
          onFilmCardMouseOut={() => {}}
          isPlaying={false}
        />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
