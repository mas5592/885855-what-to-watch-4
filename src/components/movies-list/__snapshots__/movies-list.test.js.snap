// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviesList renders correctly 1`] = `
<div
  className="catalog__movies-list"
>
  <article
    className="small-movie-card catalog__movies-card"
    id={1}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="Aviator"
        height={175}
        src="img/aviator.jpg"
        width={280}
      />
    </div>
    <h3
      className="small-movie-card__title"
      onClick={[Function]}
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        Aviator
      </a>
    </h3>
  </article>
  <article
    className="small-movie-card catalog__movies-card"
    id={2}
    onMouseOver={[Function]}
  >
    <div
      className="small-movie-card__image"
    >
      <img
        alt="The Grand Budapest Hotel"
        height={175}
        src="img/bg-the-grand-budapest-hotel.jpg"
        width={280}
      />
    </div>
    <h3
      className="small-movie-card__title"
      onClick={[Function]}
    >
      <a
        className="small-movie-card__link"
        href="movie-page.html"
      >
        The Grand Budapest Hotel
      </a>
    </h3>
  </article>
</div>
`;
